wit/types.wit

```diff
@@ -157,32 +140,74 @@ interface types {
   // type (that uses the single `stream` type mentioned above). The `consume` and
   // `write` methods may only be called once (and return failure thereafter).
-  // The `headers` and `trailers` passed into and out of responses are shared
-  // with the response, with all mutations visible to all uses. Components MUST
-  // avoid updating `headers` and `trailers` after passing a response that
-  // points to them to the outside world.
-  type incoming-response = u32;
-  type outgoing-response = u32;
-  drop-incoming-response: func(response: incoming-response);
-  drop-outgoing-response: func(response: outgoing-response);
-  incoming-response-status: func(response: incoming-response) -> status-code;
-  incoming-response-headers: func(response: incoming-response) -> headers;
-  incoming-response-consume: func(response: incoming-response) -> result<incoming-stream>;
-  new-outgoing-response: func(
-    status-code: status-code,
-    headers: headers
-  ) -> result<outgoing-response, error>;
-  outgoing-response-write: func(response: outgoing-response) -> result<outgoing-stream>;
+  resource incoming-response {
+    status: func() -> status-code;
 
-  // The following block defines a special resource type used by the
-  // `wasi:http/outgoing-handler` interface to emulate
-  // `future<result<response, error>>` in advance of Preview3. Given a
-  // `future-incoming-response`, the client can call the non-blocking `get`
-  // method to get the result if it is available. If the result is not available,
-  // the client can call `listen` to get a `pollable` that can be passed to
-  // `io.poll.poll-oneoff`.
-  type future-incoming-response = u32;
-  drop-future-incoming-response: func(f: future-incoming-response);
-  future-incoming-response-get: func(f: future-incoming-response) -> option<result<incoming-response, error>>;
-  listen-to-future-incoming-response: func(f: future-incoming-response) -> pollable;
+    headers: func() -> headers;
+
+    // May be called at most once. returns error if called additional times.
+    // TODO: make incoming-request-consume work the same way, giving a child
+    // incoming-body.
+    consume: func() -> result<incoming-body>;
+  }
+
+  resource incoming-body {
+    // returned input-stream is a child - the implementation may trap if
+    // incoming-body is dropped (or consumed by call to
+    // incoming-body-finish) before the input-stream is dropped.
+    // May be called at most once. returns error if called additional times.
+    %stream: func() -> result<input-stream>;
+
+    // takes ownership of incoming-body. this will trap if the
+    // incoming-body-stream child is still alive!
+    finish: static func(this: incoming-body) -> future-trailers;
+  }
+
+  resource future-trailers {
+    /// Pollable that resolves when the body has been fully read, and the trailers
+    /// are ready to be consumed.
+    subscribe: func() -> pollable;
+
+    /// Retrieve reference to trailers, if they are ready.
```

Comment:
It looks like you're switching to doc-comments in a few places, but not all.  It seems like we should be all-doc-comments or no-doc-comments (which is making me ask once again: why is this a valuable choice?!).

Reply:
I think it would be worthwhile to make all comments doc comments: all the comments we write now are meant to document the api interface, instead of remain as notes for future readers.